assets/blocks/course-outline/outline-block/existing-lessons-modal/use-add-existing-lessons.js

```diff
@@ -0,0 +1,94 @@
+/**
+ * WordPress dependencies
+ */
+import { createBlock } from '@wordpress/blocks';
+import { select, useDispatch, useSelect } from '@wordpress/data';
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * Lesson type difinition.
+ *
+ * @typedef {Object} Lesson
+ * @property {number} id    Lesson ID.
+ * @property {string} title Lesson title
+ */
+
+/**
+ * Find a question block based on question ID, or title if ID is missing.
+ *
+ * @param {Array}  blocks
+ * @param {Lesson} item
+ */
+export const findLessonBlock = ( blocks, { id, title } ) => {
+	const compare = ( { attributes } ) =>
+		id === attributes.id ||
+		( ! attributes.id && attributes.title && attributes.title === title );
+
+	return blocks.find( compare );
+};
+
+const API_PATH = '/sensei-internal/v1/lessons/prepare';
+
+/**
+ * Add existing lessons to the course outline block.
+ *
+ * @param {string} clientId The quiz block client id.
+ * @return {Function} Function that takes an array of lesson IDs and returns a Promise.
+ */
+export const useAddExistingLessons = ( clientId ) => {
+	const lessonBlocks = useSelect(
+		() => select( 'core/block-editor' ).getBlocks( clientId ),
+		[]
+	);
+	const { insertBlock } = useDispatch( 'core/block-editor' );
+	const nextInsertIndex = useSelect(
+		() => select( 'core/block-editor' ).getBlockCount( clientId ),
+		[]
+	);
+	const courseId = useSelect(
+		() => select( 'core/editor' ).getCurrentPostId(),
+		[]
+	);
+
+	return ( lessonIds ) => {
+		const newLessonIds = lessonIds.filter( ( lessonId ) => {
+			return (
+				lessonBlocks.length === 0 ||
+				! findLessonBlock( lessonBlocks, { id: lessonId } )
+			);
+		} );
+
+		if ( newLessonIds.length === 0 ) {
+			return Promise.resolve( {} );
+		}
+
+		// Put this before the auto-block.
+		let insertIndex = nextInsertIndex;
+
+		return apiFetch( {
```

Comment:
I briefly considered an option of doing that on the saving step.
But if I understand it correctly, it requires significant changes in other part parts of the outline. Because we need to substitute IDs when we going to duplicate a lesson. In the best-case scenario, I think, it just causes double saving with a bit glitchy title (different statuses of that lesson might be shown). On the other hand, it might be not so painful. (But honestly, I don't see an issue with duplication as well. However, it might be due my bias here).
Do you think we should re-do it that way?

Reply:
Personally, I don't like the behavior of creating the lessons before the action of publishing because I imagine users don't expect this. If they give up the lesson duplication and later they see the lesson created, it could be confusing.

But maybe we could also go with this as it is, and change it in the future if we hear complaints.

cc @donnapep to see if she has a better opinion about this from a product perspective (Notice that the lesson post is duplicated when adding to the editor, but it's associated with the course only after saving it).

Reply:
Eventually, I tried to implement it the way @renatho suggested. Here is my attempt: https://github.com/Automattic/sensei/pull/7469

Unfortunately, it causes the infinite saving loop, because the content from the server comes different then the one in the editor: IDs mismatch. And it saves it again and again, duplicating lessons again and again :) Be prepared to get a lot of duplicates while testing it :)

I'd prefer to go further with the solution from that PR instead of moving the one in this PR behind the feature flag. (Just because using the feature flag here feels like a failure for me :)) However, I didn't find a good solution how to resolve the issue with matching structure. If you have no ideas either, I'm adding a feature flag.

Reply:
As we show only courses that are not attached to a lesson, what do you think is better to display here? I mean how to communicate that we allow to add only "courseless" lessons?

![CleanShot 2024-02-07 at 04 45 24@2x](https://github.com/Automattic/sensei/assets/329356/d22cbfd0-e719-4128-b950-d4ceab107433)

Reply:
"Courseless" isn't used much so it's a bit strange. 😄 "Available" may be a better choice.

Reply:
Updated here: https://github.com/Automattic/sensei/pull/7368/commits/0b51e794cbc059e13696f31c6baf6cca2f7a0f47

![CleanShot 2024-02-08 at 14 48 09@2x](https://github.com/Automattic/sensei/assets/329356/f4b16d8e-8338-44e7-b695-913e3ec5f96b)